File: src/typescript/fable-core/AsyncBuilder.ts
import { IDisposable } from "./Util"

export type Continuation<T> = (x: T) => void;

export interface CancellationToken {
  isCancelled: boolean;
}

export class Trampoline {
  static get maxTrampolineCallCount() {
    return 2000;
  }
  private callCount: number;
  constructor() {
    this.callCount = 0;
  }
  incrementAndCheck() {
    return this.callCount++ > Trampoline.maxTrampolineCallCount;
  }
  hijack(f:() => void) {
    this.callCount = 0;
    setTimeout(f, 0);
  }
}

export interface IAsyncContext<T> {
  onSuccess: Continuation<T>;
  onError: Continuation<any>;
  onCancel: Continuation<any>;

  cancelToken: CancellationToken;
  trampoline: Trampoline;
}

export type IAsync<T> = (x: IAsyncContext<T>) => void;

export function protectedCont<T>(f: IAsync<T>) {
    return (ctx: IAsyncContext<T>) => {
        if (ctx.cancelToken.isCancelled)
            ctx.onCancel("cancelled");
        else if (ctx.trampoline.incrementAndCheck())
            ctx.trampoline.hijack(() => {
                try {
                    f(ctx);
                } catch (err) {
                    ctx.onError(err);
                }
            });
        else
            try {
                f(ctx);
            } catch (err) {
                ctx.onError(err);
            }
    };
}

export function protectedBind<T, U>(computation: IAsync<T>, binder: (x: T) => IAsync<U>) {
    return protectedCont((ctx: IAsyncContext<U>) => {
        computation({
            onSuccess: (x: T) => {
              try {
                binder(x)(ctx)
              }
              catch (ex) {
                ctx.onError(ex);
              }
            },
            onError: ctx.onError,
            onCancel: ctx.onCancel,
            cancelToken: ctx.cancelToken,
            trampoline: ctx.trampoline
        });
    });
}

export function protectedReturn<T>(value?: T) {
    return protectedCont((ctx: IAsyncContext<T>) => ctx.onSuccess(value));
}

export class AsyncBuilder {
  Bind<T, U>(computation: IAsync<T>, binder: (x: T) => IAsync<U>) {
    return protectedBind(computation, binder);
  }

  Combine<T>(computation1: IAsync<void>, computation2: IAsync<T>) {
    return this.Bind(computation1, () => computation2);
  }

  Delay<T>(generator: () => IAsync<T>) {
    return protectedCont((ctx: IAsyncContext<T>) => generator()(ctx));
  }

  For<T>(sequence: Iterable<T>, body: (x: T) => IAsync<void>) {
    const iter = sequence[Symbol.iterator]();
    let cur = iter.next();
    return this.While(() => !cur.done, this.Delay(() => {
      const res = body(cur.value);
      cur = iter.next();
      return res;
    }));
  }

  Return<T>(value?: T) {
    return protectedReturn(value);
  }

  ReturnFrom<T>(computation: IAsync<T>) {
    return computation;
  }

  TryFinally<T>(computation: IAsync<T>, compensation: () => void) {
    return protectedCont((ctx: IAsyncContext<T>) => {
      computation({
        onSuccess: (x: T) => {
          compensation();
          ctx.onSuccess(x);
        },
        onError: (x: any) => {
          compensation();
          ctx.onError(x);
        },
        onCancel: (x: any) => {
          compensation();
          ctx.onCancel(x);
        },
        cancelToken: ctx.cancelToken,
        trampoline: ctx.trampoline
      });
    });
  }

  TryWith<T>(computation: IAsync<T>, catchHandler: (e: any) => IAsync<T>) {
    return protectedCont((ctx: IAsyncContext<T>) => {
      computation({
        onSuccess: ctx.onSuccess,
        onCancel: ctx.onCancel,
        cancelToken: ctx.cancelToken,
        trampoline: ctx.trampoline,
        onError: (ex: any) => {
          try {
            catchHandler(ex)(ctx)
          }
          catch (ex2) {
            ctx.onError(ex2);
          }
        }
      });
    });
  }

  Using<T extends IDisposable, U>(resource: T, binder: (x: T) => IAsync<U>) {
    return this.TryFinally(binder(resource), () => resource.Dispose());
  }

  While(guard: () => boolean, computation: IAsync<void>): IAsync<void> {
    if (guard())
      return this.Bind(computation, () => this.While(guard, computation));
    else
      return this.Return(void 0);
  }

  Zero() {
    return protectedCont((ctx: IAsyncContext<void>) => ctx.onSuccess(void 0));
  }
}

export const singleton = new AsyncBuilder();

File: src/fable-library-py/fable/asyncbuilder.py
from typing import Any, Callable, Optional, TypeVar, Awaitable
from expression.core.aiotools import from_result

T = TypeVar("T")
U = TypeVar("U")


class AsyncBuilder:
    def Bind(self, computation: Awaitable[T], binder: Callable[[T], Awaitable[U]]) -> Awaitable[U]:
        async def bind() -> U:
            t = await computation
            return await binder(t)

        return bind()

    def Combine(self, computation1: Awaitable[None], computation2: Awaitable[T]) -> Awaitable[T]:
        return self.Bind(computation1, lambda _: computation2)

    def Delay(self, generator: Callable[[], Awaitable[T]]) -> Awaitable[T]:
        async def deferred() -> T:
            return await generator()

        return deferred()

    def Return(self, value: Optional[T] = None) -> Awaitable[Optional[T]]:
        return from_result(value)

    def ReturnFrom(self, computation: Awaitable[T]) -> Awaitable[T]:
        return computation

    def TryFinally(self, computation: Awaitable[T], compensation: Callable[[], None]) -> Awaitable[T]:
        async def try_finally() -> T:
            try:
                t = await computation
            finally:
                compensation()
            return t

        return try_finally()

    def TryWith(self, computation: Awaitable[T], catchHandler: Callable[[Any], Awaitable[T]]) -> Awaitable[T]:
        async def try_with() -> T:
            try:
                t = await computation
            except Exception as exn:
                t = await catchHandler(exn)
            return t

        return try_with()

    def Using(self, resource: T, binder: Callable[[T], Awaitable[U]]) -> Awaitable[U]:
        return self.TryFinally(binder(resource), lambda: resource.Dispose())

    def While(self, guard: Callable[[], bool], computation: Awaitable[None]) -> Awaitable[None]:
        if guard():
            return self.Bind(computation, lambda _: self.While(guard, computation))
        else:
            return self.Return()

    def Zero(self) -> Awaitable[None]:
        return from_result(None)


singleton = AsyncBuilder()

__all__ = ["singleton"]
